### ISSUE TITLE
AttributedString with CRLF sequence crashes while accessing CharacterView

### ISSUE THREAD

**Issue by chockenberry** [apple/swift#61532]
**Describe the bug**
An AttributedString with a CRLF ("\r\n") sequence will crash with a string index out of bounds if you try to access the characters.

**Steps To Reproduce**
```
import Foundation

let okString = "WTF\r\nHELLO"
let wtfString = AttributedString(okString)

if let lineRange = okString.range(of: "\n") {
	let lineSubstring = okString[okString.startIndex..<lineRange.lowerBound]
	type(of: lineSubstring)
	lineSubstring.count
}

if let lineRange = wtfString.range(of: "\n") {
	let lineSubstring = wtfString[wtfString.startIndex..<lineRange.lowerBound]
	type(of: lineSubstring.characters)
	lineSubstring.characters.first!
	lineSubstring.characters.last!
	lineSubstring.characters.count // BOOM
	// same with lineSubstring.characters.underestimatedCount
}
```
The code above will crash on line 17 (BOOM)

**Expected behavior**
The count should be the same as the unattributed string: 4

**Screenshots**
N/A

**Environment (please fill out the following information)**
 - macOS 12.5.1
 - Xcode 14.0 (14A309)

**Additional context**
Note that the last Character in the AttributedString.CharacterView is "F", not "\r".

**Comment by chockenberry** [comment]
This issue also happens Xcode 14.1 beta 3 (14B5033e) on macOS 12.5.1.

**Comment by ole** [comment]
I'm seeing not only the crash in line 17, but also an EXC_BAD_ACCESS crash in line 9 (the `lineSubstring.count` line in the first if block). Tested in an Xcode playground in Xcode 14.0.1 on macOS 12.6.

<img width="1023" alt="Screen Shot 2022-10-10 at 22 48 20" src="https://user-images.githubusercontent.com/5010/194951558-29706ba4-a6ae-44c5-9fa6-90f0df9d3238.png">

So I don't think this crash has anything to with `AttributedString`, but rather caused by the interplay between `String` and `NSString` semantics. CRLF is a single grapheme cluster in Unicode = a single `Character` in Swift. So by `String` semantics, `okString.range(of: "\n")` should *not* find a match because the string doesn’t contain a single "\n" Character. But: `range(of:)` is an `NSString` API and thus uses `NSString` semantics, where "\r\n" are two unichars. So it *does* find a match.

However, the returned `Range<String.Index>` is not valid for accessing the string in terms of Characters because it doesn’t fall on a `Character` boundary. Hence the crash.

**Comment by chockenberry** [comment]
Ole's observation makes sense. Changing the code from `lineSubstring.characters.count` to `lineSubstring.unicodeScalars.count` works around the issue.

**Comment by chockenberry** [comment]
So...

I've been looking at ways to avoid `.range(of:)` in this code and it's pretty ugly for AttributedStrings.

The issue is that all of the current mechanisms that rely on the character set of grapheme clusters, such as `.isNewline`, all return a `String.Index`. You can use this information to find a newline boundary with something like this:

```
var substring = okString[okString.startIndex..<okString.endIndex]
if let newlineIndex = substring.firstIndex(where: { $0.isNewline }) {
    let foundString = substring[..<newlineIndex]
    ...
}
```

But that `newlineIndex` is a `String.Index` and you need an `AttributedString.Index`. You're now faced with doing:

```
    if let startIndex = AttributedString.Index(foundString.startIndex, within: wtfString),
       let endIndex = AttributedString.Index(foundString.endIndex, within: wtfString) {
        print("wtfString = \(wtfString[startIndex..<endIndex])")
    }
```

That's a lot of boilerplate to split a string.

And I'm not entirely sure it's a good idea to assume the indices of a `String` will be the same in an `AttributedString`. Maybe the `AttributedString.Index` initializer is protecting me from the implementation details of the attributed, but so would being able to use `firstIndex`, `split`, et. al. on an attributed string instance. And it would certainly be a lot easier to read.

**Comment by ole** [comment]
> The issue is that all of the current mechanisms that rely on the character set of grapheme clusters, such as `.isNewline`, all return a `String.Index`.

You can use the `characters` view on `AttributedString` for this. It is a `Collection` of `Character`s, so has a similar API as `String`, but works in terms of `AttributedString.Index`. Example:

```swift
import Foundation

let okString = "WTF\r\nHELLO"
var attrString = AttributedString(okString)

if let newline = attrString.characters.firstIndex(where: { $0.isNewline }) {
    let foundString = attrString.characters[..<newline]
    print(foundString.count)
    // Or use the range on the attributed string to manipulate attributes
    attrString[..<newline].link = URL(string: "https://example.com")
}
```

**Comment by ole** [comment]
> And that an `AttributedString.CharacterView` is not the same thing as a `Substring.CharacterView`.

`String.CharacterView` and `Substring.CharacterView` are remnants of the Swift 3 era when the `String` type itself wasn't a Collection — you always had to go through one of the string views, one of which was the `Character`s view. Since Swift 4.0 they're no longer in use (and defined as typealiases to their parent types).

`AttributedString.CharacterView` has to be a different type than a plain `String` (or `Substring`) because it uses a different index type. (There are probably more reasons, but this is one.) I don't know why `AttributedString` uses a different index type than `String` in the first place, though. It looks like it's just a wrapper around a `String.Index`, so it doesn't store additional information:

```swift
let attrString = AttributedString("Hello")
dump(attrString.startIndex)
// Prints:
// ▿ Foundation.AttributedString.Index
//   ▿ characterIndex: Swift.String.Index
//     - _rawBits: 1
```

Maybe the authors of `AttributedString` introduced it to prevent index sharing between `String` and `AttributedString` values (as you noted above, sharing indices between collections is generally not safe).

> That's super helpful. The problem for me is that I saw `characters` as a `CharacterView`. I didn't make the connection that its element is a `Character` (nor did the documentation help by pointing this out).

Yeah. You have to click through to `AttributedString.CharacterView.Element` to see that the collection's element type is `Character`. I agree that the documentation could present this better.

> As someone who's decidedly not an expert with Swift, it feels really odd that there are two separate signatures for doing the same thing with a `Substring` and `AttributedString`:
> 
> ```
> func firstIndex(where predicate: (Self.Element) throws -> Bool) rethrows -> Self.Index?
> func firstIndex(where predicate: (Character) throws -> Bool) rethrows -> AttributedString.Index?
> ```

The different return types make sense, but it’s strange that one of the methods uses `Character` and one uses `Self.Element`, despite these being one and the same. I presume the documentation generator just replicates how the API authors wrote the declarations.

**Comment by lorentey** [comment]
Quick note: in Swift 5.6 and below, `Substring` used to have serious problems when its start/endIndex happened to fall in between `Character` boundaries, like in the `okString` case above. (Some of which are detailed in [a forum thread](https://forums.swift.org/t/string-index-unification-vs-bidirectionalcollection-requirements/55946), although the [associated PR](https://github.com/apple/swift/pull/41417) is probably a better summary of the issues.)

The case of the `okString` code below was one of the instances that got fixed in Swift 5.7 -- this code used to (incorrectly) trap in `count`, too:

```swift
if let lineRange = okString.range(of: "\n") {
	let lineSubstring = okString[okString.startIndex..<lineRange.lowerBound]
	type(of: lineSubstring)
	lineSubstring.count // Swift/StringCharacterView.swift:60: Fatal error: String index is out of bounds
}
```

At first glance, some of the underlying issues are probably still there in `AttributedString`, and will need a separate fix.

**Comment by ole** [comment]
@lorentey Thanks for this.

@chockenberry Did you run your tests on the iOS simulator (i.e. in a Swift 5.7 environment)? If yes, this would explain the differences in behavior you and I observed. I tested on macOS 12.6 for a macOS target (Xcode playground), so in a Swiff 5.6 environment (despite using the Swift 5.7 compiler).

**Comment by chockenberry** [comment]
@ole Yep, that's exactly what happened. I just created a new playground with macOS (Swift 5.6) and see the same "String index is out of bounds" that you see.
